File: Problems/Leetcode/Java/LC142LinkedListCycleII.java
import java.util.Hashtable;

class ListNode {
    int val;
    ListNode next;

    ListNode(int x) {
        val = x;
        next = null;
    }
}

public class LC142LinkedListCycleII {

    public static ListNode detectCycleOptimized(ListNode head) {
        if (head == null || head.next == null) {
            return null;
        }

        ListNode slow = head, fast = head, entry = head;

        while (fast.next != null && fast.next.next != null) {
            slow = slow.next;
            fast = fast.next.next;

            if (fast == slow) {
                while (entry != slow) {
                    entry = entry.next;
                    slow = slow.next;
                }
                return entry;
            }
        }

        return null;
    }

    public static ListNode detectCycle(ListNode head) {
        if (head == null) {
            return null;
        }

        ListNode node = head;
        Hashtable<ListNode, Integer> h = new Hashtable<>();

        int index = 0;
        while (node != null) {
            int i = h.getOrDefault(node, -1);

            if (i == -1) {
                h.put(node, index);
            } else {
                return node;
            }

            node = node.next;
            index++;
        }

        return null;
    }

    public static void main(String[] args) {
        ListNode head = new ListNode(1);
        ListNode node1 = new ListNode(2);
        ListNode node2 = new ListNode(3);
        ListNode node3 = new ListNode(4);
        ListNode node4 = new ListNode(5);
        ListNode node5 = new ListNode(6);
        ListNode node6 = new ListNode(7);

        head.next = node1;
        node1.next = node2;
        node2.next = node3;
        node3.next = node4;
        node4.next = node5;
        node5.next = node6;
        node6.next = node3;

        System.out.println(detectCycleOptimized(head));
    }
}

File: Problems/Leetcode/Python/LC142LinkedListCycleII.py
from typing import List, Optional


class Node():
    def __init__(self, data) -> None:
        self.data = data
        self.next = None


def detectCycle(head: Optional[Node]) -> Optional[Node]:
    if head is None:
        return None

    is_crossed_at = {}
    node = head

    index = 0
    while node:
        ind = is_crossed_at.get(node, -1)

        if ind != -1:
            return node
        else:
            is_crossed_at[node] = index

        node = node.next
        index += 1

    return None


def createLL(arr: List[int]):
    if len(arr) == 0:
        return None
    head = Node(arr[0])
    last = head
    for data in arr[1:]:
        last.next = Node(data)
        last = last.next
    return head


def printLL(head: Node):
    while head:
        print(head.data, end=" ")
        head = head.next
    print()


# arr = [int(i) for i in input().split()]
# head = createLL(arr)
node1 = Node(1)
node2 = Node(2)
node3 = Node(3)
node4 = Node(4)
node5 = Node(5)
node1.next = node2
node2.next = node3
node3.next = node4
node4.next = node5
node5.next = node2
res = detectCycle(None)
print(res)
